### DISCUSSION TITLE
Setting the client id dynamically

### DISCUSSION THREAD

**Discussion by rafsawicki** [emqx/emqx#13995] [General]
Hello,

Is there any way to dynamically set the client id? I imagine that the most likely place where this could be possible would be using the plugin or gRPC hook, but it wasn't able to find any examples or documentation. I've gone over the available hooks documentation, but there doesn't seem to be a way to return a new value of client id from any of them. I do not know much about erlang, so I might be missing something obvious.

It does seem that EMQX have some capacity of overriding the client id, as there's an option to replace it with an username, and I found some examples of setting the client id from the client certificate.

**Comment by id** [comment]
Hi @rafsawicki. I don't quite understand your use case. What exactly are you trying to do?

**Reply by rafsawicki** [reply]
The use case that I have, is that I'm trying to allow multiple connections from a single device to the broker. Some devices that connect to the server I'm managing are somewhat buggy and in certain scenarios, they try to establish multiple long lived connections to the broker instead of just one. Since I wasn't able to find an option to enable that (obviously out of spec, but available in some brokers) behavior in EMQX, my plan was to dynamically override the client id that was sent with a combination of the username and a random string

**Reply by rafsawicki** [reply]
It's possible that using hooks it not the best option here and I've simply searched in wrong places - my first though was to reach out for a plugin, because I've used VerneMQ before I migrated to EMQX due to performance reasons, and authentication hooks there allowed to override the client information

**Reply by id** [reply]
If you don't care about client id, set it to empty string, and EMQX will generate a random client id to identify this connection.

**Reply by id** [reply]
https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901059

> A Server MAY allow a Client to supply a ClientID that has a length of zero bytes, however if it does so the Server MUST treat this as a special case and assign a unique ClientID to that Client [MQTT-3.1.3-6]. It MUST then process the CONNECT packet as if the Client had provided that unique ClientID, and MUST return the Assigned Client Identifier in the CONNACK packet [MQTT-3.1.3-7].

**Reply by rafsawicki** [reply]
Typically that would be my first choice as well, but unfortunately, a company that developed software for those devices decided in their brilliance, that if a client id is not set (or set to an empty value), they will just assign a default static client id themselves, based on a device serial number

**Reply by id** [reply]
Ok, I see.

Yes, as you mentioned in your original post, if you use mTLS then you can map some info from client certificate to client id https://docs.emqx.com/en/emqx/latest/access-control/authn/x509.html#configure-via-dashboard.

If not, I'm not sure. May be my colleagues would be able to help.

**Reply by id** [reply]
But even with the certificate, you probably going to use one cert per device, and the information will be static again.

**Reply by rafsawicki** [reply]
As you are saying, it would still stay static if a certificate was used. I mentioned those places that I've found, as they seem to imply that EMQX has a possibility of changing the client id internally, no matter what a client has sent in the CONNECT packet. I kinda hoped that maybe there was some non-obvious internal function that someone knows of, that I could call from the plugin, that would change the client properties.

**Reply by zmstone** [reply]
Hi @rafsawicki 
Unfortunately, we cannot do this without code change.
Not with a plugin either.

**Reply by rafsawicki** [reply]
Oh, that's unfortunate to hear. I guess my only option left is to see if I can control this field by adding an additional load balancer that has capability of writing plugins with packet manipulation. Since I'm not looking forward into complicating my infrastructure setup, I hope that more advanced hook capabilities will make it one day to a feature list - being able to set more client properties the same way as MQTT 5.0 client properties or attributes can currently be set would be lovely.

Big thanks to everyone participating in the discussion.

**Reply by zmstone** [reply]
The current `client.connect` hook-point does not support mutating client ID.
And I would not recommend making an gRPC call for this (even if `client.connect` supported it).

If the client ID can be derived from other factors, such as username, a field in the decoded password token etc.
i.e. if it's purely computational, it's actually rather simple to make EMQX support it.

Today, EMQX can override clientid in two different ways:

1. use username as clientid (which is rather odd, but my guess is it was from an old paying customer)
2. the standard randomized clientid generation if the client did not provide one

https://github.com/emqx/emqx/blob/2bf8ee62368fc8c4f25a23bc382a04b72b6ee793/apps/emqx/src/emqx_channel.erl#L1661-L1662

If the clientID to be assigned is a simple enough extraction, or hashing etc. we could accept a pull request to have it supported.

**Comment by zmstone** [comment]
Planed to add a expression based clientid override rule in 5.9 (as a part of the multi-tenancy feature).

e.g.

```
## equivalent to use_username_as_clientid = true
clientid_override = "username"

## equivalent to peer_cert_as_clientid = cn
clientid_override = "cn"

## for multi-tenancy, override with a tns prefix
clientid_override = "concat(client_attrs.tns,'-',coalesce(clientid,rand_str(12)))"
```

not sure if this meets your requirements @rafsawicki , but the expression can be quite flexible, so there is a good chance you don't need a gRPC.

**Reply by rafsawicki** [reply]
Since I wanted to have a client id being derived from the username and a random string, this will definitely solve my issue. Looking forward into testing it, thanks a lot.

**Reply by zmstone** [reply]
there are a few others waiting for this feature, so i'm trying to push this to the 5.8.3 release.
https://github.com/emqx/emqx/pull/14195
